Documentation

Mingling94 opened the pull request:
Moved the plugins back into a plugins folder, amended documentation to address #18
#14 isn't completely solved because I want to approach this using JSDoc 3 standards.

https://github.com/Mingling94/test-ui/blob/982d8c2c18bc1ed7d6f8e9e69a703a179055cdc8/doc/plugin-tutorial.md

Checkout the tutorial :D

Mingling94 committed:
Added support and customization for jsdoc3

Mingling94 committed:
reorganized repo and reconfigured package.json

Mingling94 committed:
More tutorial and some webpage example plugins

Mingling94 committed:
Various fixes in code and commenting

Based on David's thoughts in issue #18, I changed the api calling to be
one at a time instead of using an array of apicalls.

Mingling94 committed:
Writing the plugin some more to include images

Mingling94 committed:
Added some images and more tutorial

Mingling94 committed:
Further tutorialing

Mingling94 committed:
Finished the tutorial

It may need some editing

DavidVorick commented:
I looked at the tutorial first, looks pretty good. I think at the end you should display the whole overview.js file so people can see it's completed form. I believe you show the entirety of the other files.

DavidVorick commented:
``` js
// Variables to store API call values
var calls = ['/wallet/status', '/gateway/status', '/consensus/status'];

function callAPI() {
    calls.forEach(function(call) {
        IPC.sendToHost('api-call', call);
    });
}
```

This portion is a bit sloppy. Namely, the word 'call' appears wayyy too many times. Also, I would think that you would want the `calls` variable to be inside of the function, is there a reason it isn't?

DavidVorick commented on app/js/daemonManager.js:
```diff
@@ -45,17 +22,13 @@ var Daemon = (function() {
-			};
+		if (!isNotRunning) {
+			isNotRunning = function() {};
 		}
```
@lukechampine might have some advice

are these checks necessary? I think it's safe to assume that the caller didn't use null.

And, since you DO use null, it's probably better to just send in an empty function.

DavidVorick replied:
I'm also inclined to say that this abstraction is unnecessary, how often do you call `ifSiad`?

Mingling94 replied:
A few times, starting and stopping uses it for safety. Otherwise, other modules might be introduced which could use this. 

Honestly it's just to save me from writing code to send a test call 3x over.

As for the null --> empty function checks, why not?

DavidVorick replied:
it has opaque usage. I'm wondering what Luke's opinion is on this.

3x is a good reason to build an extra function.

DavidVorick commented on app/js/daemonManager.js:
```diff
@@ -84,45 +57,30 @@ var Daemon = (function() {
-		APIJS.getCall(address + '/daemon/stop', function(err, data) {
+		apiCall('/daemon/stop', function(err, data) {
 			console.assert(!err && data);
 			console.log(data);
```
what happens if the daemon is already not running?

DavidVorick replied:
the user could kill it from the task manager, for example.

Mingling94 replied:
The check a few lines up using ifSiad should already prevent stopping a daemon that isn't running.

DavidVorick commented on app/js/plugin.js:
```diff
@@ -10,34 +7,20 @@ var Factory = require('./pluginFactory');
-		//view.style.display = '';
-		view.send('show');
+		view.style.display = '';
+		view.executeJavaScript('if (typeof init === "function") init();');
```
the `if` statement seems weird.

I guess it makes sense if you're trying to be paranoid about people writing plugins incorrectly.

Mingling94 replied:
Paranoid, maybe. This really just prevents error output from each plugin that lacks an init/kill function. No need to force these on the plugins, right?

DavidVorick replied:
I guess not.

DavidVorick commented on app/js/plugin.js:
```diff
@@ -56,7 +39,8 @@ module.exports = function plugin(plugPath, name) {
 	function sendIPC(channel, args) {
-		view.send(channel, args);
+		view.send.apply(view, [].slice.call(arguments));
+		//view.send(channel, args);
```
what was the reason for changing this?

Mingling94 replied:
This one's an interesting JS thing. Calling `anyFunction.apply(this_param, array)` calls the function, giving this_param as the `this` value and an array to allow a variable amount of arguments. I encountered problems sending, not just the callresult from an API call, but the original call string/object as well as a possible error. I would have to send a specific number of arguments otherwise.

In fact sendIPC shouldn't even have parameters anymore.

DavidVorick commented on app/js/pluginFactory.js:
```diff
@@ -24,16 +25,28 @@ module.exports = {
 		v.nodeintegration = 'on';
 		// TODO: If security becomes a concern, specify preload script to give
 		// plugins their dependencies and access to libraries, but turn off
 		// nodeintegration superpowers
```
neither of these TODOs is really a concern.

Mingling94 replied:
Word, I'll take them out tomorrow.

DavidVorick commented on app/plugins/Overview/index.html:
```diff
@@ -2,23 +2,31 @@
 <html>
 	<head>
 		<title>Overview</title>
```
does having a title make sense?

DavidVorick replied:
ditto for the tutorial

DavidVorick commented on app/plugins/Overview/js/overview.js:
```diff
@@ -0,0 +1,72 @@
+	calls.forEach(function(call) {
+		IPC.sendToHost('api-call', call);
+	});
+}
```
this function just seems like it could be better.

Mingling94 replied:
Alternatively, it could be individually called:

``` js
IPC.sendToHost('api-call', '/wallet/status');
IPC.sendToHost('api-call', '/gateway/status');
IPC.sendToHost('api-call', '/consensus/status');
```

Or use the standard for loop.

I could just rename the repetition of the word 'call' out of this. I'm of the opinion `array.forEach(function(element, index, array) { // Executed per array element });` is preferable to for loops in JS

DavidVorick replied:
array.forEach is an odd construction to me, I think a for loop would be more familiar to most people.

I'm not too worried about it though.

DavidVorick commented on app/plugins/Overview/js/overview.js:
```diff
@@ -0,0 +1,72 @@
+}
+
+// Update values per call
+IPC.on('api-result', update);
```
This file has no docstrings

Mingling94 replied:
Gonna do some JSDoc flexing tomorrow :tongue:

DavidVorick commented:
Ok, that sums up my review of this PR. Not too many things.

Mingling94 committed:
Various changes to make plugin writing easier

overview makes api calls not through a for loop. It now receives
results by listening on the channel corresponding to the endpoint string

the daemon uses arrays of api call parameters instead of a predefined
call JS object. The first index of a call param array is the url or
endpoint.

Removed html titles.

Added debugging node package, node-inspector, used for debugging the
main process. Try it by running `npm run debug`
